consul/datadog_checks/consul/consul.py

```diff
@@ -479,3 +503,43 @@ def check_network_latency(self, instance, agent_dc, main_tags):
                 self.gauge('consul.net.node.latency.p95', latencies[ceili(n * 0.95) - 1], hostname=node_name, tags=main_tags)
                 self.gauge('consul.net.node.latency.p99', latencies[ceili(n * 0.99) - 1], hostname=node_name, tags=main_tags)
                 self.gauge('consul.net.node.latency.max', latencies[-1], hostname=node_name, tags=main_tags)
+
+    def submit_node_service_check(self, instance, instance_state, agent_members, service_check_tags):
+        # Add all nodes we see to the available list of Nodes:
+        # Health check for the Consul Agent of each node
+        # Keep track of nodes we have seen and report Critical if they are missing now or if they are failing
+        node_level_service_checks = instance.get('node_level_service_checks', False)
+        node_regexes = instance.get("node_regexes", [])
+
+        if not _is_affirmative(node_level_service_checks):
+            return
+
+        # Recompute the list of nodes based on regex parameters
+        if len(node_regexes) > 0:
+            nodes = self._filter_nodes(agent_members, node_regexes)
```

Comment:
you shouldn't filter at every iteration. you should keep a set of matching nodes and a set of non matching nodes and refilters only nodes that are not part of these sets.